### ISSUE TITLE
data could be collected into a CSV file

### ISSUE THREAD

**Issue by teslasmoustache** [fabacab/wifi-cafe-surveyor#1]
which can be analysed in several ways, including turning it into a spreadsheet and creating graphs from it.

Just need to figure out all the data that needs collecting.

**Comment by fabacab** [comment]
Again, I think we're talking at cross-purposes. CSV is a useful _reporting_ format but a terrible _analytics_ format. Especially considering we have bettrler options, even just an HTTP POST to a Google Spreadsheet makes more sense than storing directly to CSV?

**Comment by teslasmoustache** [comment]
I'm just saying, a CSV can be _converted_ into a spreadsheet very easily and analytics could be done on that. That could be done locally, or changes could be uploaded somewhere else. I'm not familiar with Google's spreadsheet tool, as I make a strong effort to avoid that company. I would prefer not to create a Google account. And I'm not familiar with the concept of HTTP POST.

I think it might be worth mentioning at this point that I'm pretty sure your kung fu is stronger than mine. As evidenced by [this terrible script I just wrote.](https://github.com/teslasmoustache/SecurityPoo)

**Comment by fabacab** [comment]
That's not such a bad script, it's just simplistic. Maybe my fu is stronger than yours, I'm not sure. Either way, I'm happy to work on a project like this with someone who's interested in working on a project like this, regardless of their fu's strength. I'm _not_ interested in working on a project like this _alone_, though.

An HTTP POST is like a Web form submission. See the [HTTP method definitions](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html), sometimes called "HTTP verbs." It's what you do with [`wget --post-data`](https://www.gnu.org/software/wget/manual/html_node/HTTP-Options.html), for example, and is a way to send data to a server (or any other service that can consume incoming HTTP requests). Same kind of idea as a query string in a URL (which would be an HTTP GET), but easier to send more data in one network connection.

What I'm trying to work out with this discussion is what the answers to the implicit questions posed in the [database information](https://github.com/meitar/wifi-cafe-surveyor/wiki/Database-information) page might be. In other words, let's assume we have a script that can collect all the information we need. So you run it on your laptop while at a café and now you have a bunch of information about that specific wifi hotspot.

Great. Now what? Where should that information go? Obviously, it does us no good if it stays on your laptop, unless your laptop travels to every wifi hotspot we're interested in knowing about. Since the point of this project is to survey every hotspot in the world (or as many hotspots as we can), surveying them from a single laptop makes no sense. That means we need to store the information collected by running the script on your laptop somewhere else. Where should that be? It should _not_ be a CSV file.

Make sense?

**Comment by teslasmoustache** [comment]
Alright. Good. I am interested in working on this.

I understand now why you would like to use a Google form/spreadsheet. I guess I didn't realize the scale you wanted to do this at. And I'm totally cool with this becoming a larger thing. I started looking for alternatives, but I'm doubtful of how well they would work for our purposes. Ethercalc is a thing that exists....?

I mean, I suppose a form could be built from scratch, but I have no idea how to do that. There's also the issue of hosting. If this were something very small, I could do it on one of my own VPSs. I don't think that's really the case here.

I'll see if I can create a Google account without giving them too much personal information.

**Comment by fabacab** [comment]
> I guess I didn't realize the scale you wanted to do this at.

To clarify, I don't actually expect to survey every hotspot in the world. But my point is that thinking at that scale and then scaling back is a much better approach than thinking at a smaller scale and then scaling up. The latter will run into trouble very quickly while the former will give us a lot of breathing room very early on. It is, IME, always better to start with breathing room than without it. :)

> I understand now why you would like to use a Google form/spreadsheet.

So I don't actually want to use a Google Form _per se_, it was just an example that gives us some "free" functionality. "Free" in quotes because, obviously, fuck Google. There are alternatives to a Google Spreadsheet that give us the same or even more features, such as designing a simple RDBMS schema ourselves. That's what I've already started doing at the [Data to collect](https://github.com/meitar/wifi-cafe-surveyor/wiki/Data-to-collect) page. These are descriptions of data that map relatively easily to tables in a SQL database.

> I suppose a form could be built from scratch, but I have no idea how to do that.

Ah, this is probably where our communication breakdown happened. A form like this is trivial to create, and I'll happily show you how it's done.

> There's also the issue of hosting. If this were something very small, I could do it on one of my own VPSs. I don't think that's really the case here.

Well, it certainly is small for now. :) And even so, I think you'll be surprised how much power even a low-end VPS can give us for a project like this.

> I'll see if I can create a Google account without giving them too much personal information.

Maybe hold off on that? What if we whip up a really quick script that accepts an HTTP POST so you can see how that's done?

On that note, what programming languages do you know? Your GitHub activity seems to show Ruby and JavaScript as the primary languages (other than shell scripts). I might enjoy building this in Node.js, but I also would enjoy brushing up on my Ruby. I'm _most_ comfortable with either Python or PHP, though, so I don't have a strong preference. Do you?

**Comment by teslasmoustache** [comment]
- Scaling

Ability to scale up sounds prudent. Glad you're on board with not using Google.
- VPS

I guess I'm alright with this being used on my server. I just worry a little about something drastic happening. But that's what backups are for.
- Programming Languages

Shell's really all I know. The Ruby you see there is Diaspora*, where I had some high hopes for something I don't even remember what. No code contributed. The JS you see is me trying to hack on a FirefoxOS app called "isitgonnarain", which would tell you, basically, whether or not it's going to rain. Very basic weather info. My version of the app would have been a bit more bitter about it than the original, reflecting my usual feelings about weather. i.e. "It's fucking snowing" rather than "you should wear snow boots! ;)"

Anyway, I'm used to figuring out Shell, and can read and  poke at Python (which I'd like to learn more of). I'm even less familiar with PHP, but I've edited a couple things here and there and it doesn't seem all that bad. Honestly, every time I look at Javascript I'm like "what the fuck is this."
- Web forms

I looked up web forms, and holy shit. Yeah. That's easy. I think there's a way to autofill data from a spreadsheet (or other file). So that should make it easier to submit the data. Is this something you've figured out already?

And...Regarding what data needs to be collected, I'm still not sure how to identify a security appliance. Gateway model can be determined from the MAC address (reliable?).

Still murky about databases, but I'm confident I can figure it out.

**Comment by teslasmoustache** [comment]
To be clear, I know we're avoiding collecting all the data into one file. I just mean the script has to export to something (even if stdio) and we use _that_ to fill out the web form.

And I'm assuming that a human has to fill some of this out? I don't think most computers have GPS modules in them, or know who the owner is, so....

**Comment by teslasmoustache** [comment]
I've got a HTML form written, and I'm cleaning up some CSS for it. Should have it tonight or tomorrow. Gotta go to work in the meantime.

**Comment by fabacab** [comment]
It's good that you're learning about forms. But…

> fill out the web form.

…we don't need a form at all. Forms are just an interface for humans. A script can skip the form entirely and send data directly to the form's handler in a machine-readable format. That's an API. The original point of scripting this was so that no human needed to type anything.

But for the sake of learning, go ahead and make a form and when you've done that share the code and I'll show you why the form part is unnecessary.

**Comment by teslasmoustache** [comment]
But a human needs to enter venue name, owner, stuff like that. Unless you want to skip that.

**Comment by fabacab** [comment]
No, a human doesn't. That's already written down. What do you think Google Maps is?

**Comment by teslasmoustache** [comment]
K...I'll be interested to see how we do that.

OK. Well, here it is. https://github.com/teslasmoustache/wifi-cafe-surveyor

**Comment by fabacab** [comment]
> OK. Well, here it is. https://github.com/teslasmoustache/wifi-cafe-surveyor

Okay, so, this is a separate repo. Are you not familiar with GitHub's "fork and pull" collaborative development model? If not, try reading some [help articles](https://help.github.com/articles/using-pull-requests/) about it.

I also gave you direct push access to this repository, so instead of making your own repository, it might have been easier to just push to this one.

I've gotta go to a thing but will have a look at the code later today.

**Comment by teslasmoustache** [comment]
yeah, I was unable to fork for some reason. I think it's because we have an empty repo here. Anyway, I'm assuming a lot of this is scratch since we don't need all this front-end stuff.

**Comment by fabacab** [comment]
Okay, gonna close this issue and move to code review to discuss details of c98e2652516f18bb6928272ab0310a34b69f307f with you and show why we don't need a UI.

**Comment by teslasmoustache** [comment]
Sure. Just thought that if you wanted location, venue, and owner information, there would have to be human intervention. Because if I use any public wifi, Google maps is accurate somewhere within a block, and OSM is accurate somewhere within like 3 blocks. For me, anyway. If the SSID is nondescript (something like "Guest Wi-fi") then no one would be able to tell.

But I guess from a statistical standpoint it wouldn't matter. You get a general area and that's good enough to observe a correlation between hardware and geographical location. That works. But with correlation between hardware and _company_ I feel like that's a slight problem. Maybe entry of this information could be optional? Anyway, that's just a whole different thing.

**Comment by fabacab** [comment]
> Just thought that if you wanted location, venue, and owner information, there would have to be human intervention. Because if I use any public wifi, Google maps is accurate somewhere within a block, and OSM is accurate somewhere within like 3 blocks. For me, anyway. If the SSID is nondescript (something like "Guest Wi-fi") then no one would be able to tell.

No, again, we can probably get company info from Wi-Fi for several reasons:
1. The code example (for the sake of example) uses GeoIP info, but we probably actually want to ask for _wifi_ location information from a database like [Skyhook](http://www.skyhookwireless.com/submit-access-point) instead.
2. Even if we can't get direct company information, once we have a statistically significant sample of wi-fi configs, we can correlate that data with the info in [Google Places](https://developers.google.com/places/) and determine which unnamed venues we should name what.
3. We can also just use other databases like Foursquare and scrape venue information.

Basically, there are a lot of approaches, all of them better than manually inputting venue names, because that's just wacky given the fact that we have computers for this shit. ;)

**Comment by fabacab** [comment]
I guess the real point I'm trying to drive home here is that if your fu is weak, you will expect that a thing cannot be done and so "obviously" we have to resort to human data entry. This is basically never the case, and you should train yourself out of thinking that it is.
